contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,1994 @@
+    let config = CONFIG.load(deps.storage)?;
+    let transfer_msg = get_transfer_msg(info.sender, rewards, config.reward_token)?;
+    Ok(Response::new()
+        .add_message(transfer_msg)
+        .add_attribute("action", "claim")
+        .add_attribute("amount", rewards))
+}
+
+pub fn get_transfer_msg(recipient: Addr, amount: Uint128, denom: Denom) -> StdResult<CosmosMsg> {
+    match denom {
+        Denom::Native(denom) => Ok(BankMsg::Send {
+            to_address: recipient.into_string(),
+            amount: vec![Coin { denom, amount }],
+        }
+        .into()),
+        Denom::Cw20(addr) => {
+            let cw20_msg = to_binary(&cw20::Cw20ExecuteMsg::Transfer {
+                recipient: recipient.into_string(),
+                amount,
+            })?;
+            Ok(WasmMsg::Execute {
+                contract_addr: addr.into_string(),
+                msg: cw20_msg,
+                funds: vec![],
+            }
+            .into())
+        }
+    }
+}
+
+pub fn update_rewards(deps: &mut DepsMut, env: &Env, addr: &Addr) -> StdResult<()> {
+    let config = CONFIG.load(deps.storage)?;
+    let reward_per_token = get_reward_per_token(deps.as_ref(), env, &config.staking_contract)?;
+    REWARD_PER_TOKEN.save(deps.storage, &reward_per_token)?;
+
+    let earned_rewards = get_rewards_earned(
+        deps.as_ref(),
+        env,
+        addr,
+        reward_per_token,
+        &config.staking_contract,
+    )?;
+    PENDING_REWARDS.update::<_, StdError>(deps.storage, addr.clone(), |r| {
+        Ok(r.unwrap_or_default() + earned_rewards)
+    })?;
+
+    USER_REWARD_PER_TOKEN.save(deps.storage, addr.clone(), &reward_per_token)?;
+    LAST_UPDATE_BLOCK.save(deps.storage, &env.block.height)?;
+    Ok(())
+}
+
+pub fn get_reward_per_token(deps: Deps, env: &Env, staking_contract: &Addr) -> StdResult<Uint256> {
+    let reward_config = REWARD_CONFIG.load(deps.storage)?;
+    let total_staked = get_total_staked(deps, staking_contract)?;
+    let current_block = min(env.block.height, reward_config.period_finish);
+    let last_update_block = LAST_UPDATE_BLOCK.load(deps.storage).unwrap_or_default();
+    let prev_reward_per_token = REWARD_PER_TOKEN.load(deps.storage).unwrap_or_default();
+    let additional_reward_per_token = if total_staked == Uint128::zero() {
+        Uint256::zero()
+    } else {
+        let numerator = reward_config
+            .reward_rate
+            .full_mul(Uint128::from(current_block - last_update_block))
+            .full_mul(scale_factor());
+        let denominator = Uint512::from(total_staked);
+        let result = numerator.checked_div(denominator)?;
+
+        // try_into() shouldn't cause problems here because final results fits into Uint256
+        result
+            .checked_div(Uint512::from(scale_factor()))?
+            .try_into()
+            .unwrap()
+    };
+
+    Ok(prev_reward_per_token + additional_reward_per_token)
```

Comment:
If I am hitting this line because I have added new funds to the contract after the previous funding set has expired should the reward rate really be the new one + the old one? Seems to me that in that case the old one shouldn't matter because it has already been paid out.

Reply:
In other words, this line seems to imply to me that the rewards rate will only ever increase. The ability to fund a rewards contract after it has completed a round of rewards seems to imply to me that is not the case as a new funding round should set a new `rewards_per_token` rate separate from the one that completed.

Reply:
yes hit is true, it is a montonically increasing value